### PR TITLE
New template and better plots

### PR DESCRIPTION
**Pull request by KevinMenden**
This PR updates the template and includes improvement to the plots generated for visualisation of the results.

### CODE REVIEW

**Comment by KevinMenden** [comment]
@Zethson in order to make this repository reproducible, I had to add a bit of code to disable the deterministic settings (which is a bit ironic) - so the linting will always fail.
Otherwise it works so - can you sleep at night if I merge this?

To a bit more I added a command line argument to disable the GPU determinism so that one can run the shell script and generate all the results.

**Comment by Zethson** [comment]
I'll have a look at it soon ;)
Maybe we can find a way to make the linter happy.

Funfact: you should have gotten an angry bit response, but got saved from it since your branch is called main and not master ;)
Something to fix @Imipenem

**Comment by KevinMenden** [comment]
Okay I tricked the linter 😈

File: .github/workflows/master_branch_protection.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+name: PR to master branch from patch/release branch only
+
+on:
+  pull_request:
+    branches:
+    - master
```
Comment (Zethson) [inline .github/workflows/master_branch_protection.yml:6]: @Imipenem we gotta figure out a solution for this. Should def run on main!

File: LICENSE [hunk 2]
```diff
@@ -1,6 +1,6 @@
 MIT License
 
-Copyright (c) 2020, Lukas Heumos
+Copyright (c) 2021, Lukas Heumos
```
Comment (Zethson) [inline LICENSE]:
```suggestion
Copyright (c) 2021, Kevin Menden, Lukas Heumos
```

File: docs/authors.rst [hunk 3]
```diff
@@ -1 +1,13 @@
-.. include:: ../AUTHORS.rst
+=======
+Credits
+=======
+
+Development Lead
+----------------
+
```
Comment (Zethson) [inline docs/authors.rst]:
```suggestion
* Kevin Menden <kevin.menden@qbic.uni-tuebingen.de>
```

File: docs/conf.py [hunk 4]
```diff
@@ -45,7 +45,7 @@
 
 # General information about the project.
 project = 'sc-autoencoder'
-copyright = "2020, Lukas Heumos"
+copyright = "2021, Lukas Heumos"
```
Comment (Zethson) [inline docs/conf.py]:
```suggestion
copyright = "2021, Kevin Menden, Lukas Heumos"
```

File: environment.yml [hunk 5]
```diff
@@ -27,9 +20,8 @@ dependencies:
...
-      - pycuda==2019.1.2  # not on Conda
-      - system-intelligence==1.2.4  # not yet on Conda
-      - scanpy==1.6.0
-      
\ No newline at end of file
+          - tensorflow==2.4.1 # not yet on Conda with full GPU support
+          - tensorflow_datasets==4.2.0
+          - pycuda==2019.1.2 # not on Conda
+          - system-intelligence==2.0.2 # not yet on Conda
+          - scanpy==1.6.0
```
Comment (Zethson) [inline environment.yml]: https://anaconda.org/bioconda/scanpy

Let's use Conda when we can

File: mlf_core.cfg [hunk 6]
```diff
@@ -2,7 +2,7 @@
 current_version = 0.1.0-SNAPSHOT
```
Comment (Zethson) [inline mlf_core.cfg]: Let's bump this to 0.1.0  + update the changelog afterwards + then we can also create the release afterwards

**Review by Zethson — COMMENTED** [review]
Don't forget to attribute yourself :)

Let's make this release ready.

What's up with the container building?

> Pip subprocess error:
> ERROR: Could not find a version that satisfies the requirement anndata>=0.7 (from scanpy)
> ERROR: No matching distribution found for anndata>=0.7

File: .github/workflows/master_branch_protection.yml (hunk 1 above)
Comment (Imipenem) [reply .github/workflows/master_branch_protection.yml:6]: @Zethson according to GH actions docs this shouldn't be more complicated than simply adding `- main`.

I will check this out tomorrow!

Comment (Zethson) [reply .github/workflows/master_branch_protection.yml:6]: Still wish there was a "default" branch, but oh well

**Comment by KevinMenden** [comment]
> What's up with the container building?

The container is building now.
It's _only_ the training that doesn't work 😏 

You've seen that one before?

```
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/docker/models/images.py", line 287, in build
    raise BuildError(chunk['error'], result_stream)
docker.errors.BuildError: name unknown
Error: Process completed with exit code 1.
```

**Comment by Zethson** [comment]
> > What's up with the container building?
> 
> The container is building now.
> It's _only_ the training that doesn't work smirk
> 
> You've seen that one before?
> 
> ```
>   File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/site-packages/docker/models/images.py", line 287, in build
>     raise BuildError(chunk['error'], result_stream)
> docker.errors.BuildError: name unknown
> Error: Process completed with exit code 1.
> ```

No, never. But I know what's wrong. We forgot to control the version in the workflow with bump-version.

The MLproject specifies version 1.0.0, but the workflow builds a container with the tag : 0.1.0-SNAPSHOT and then MLflow gets all confused.

Solution: Manually increase the version in the train_cpu.yml workflow to 1.0.0 

We'll fix this for mlf-core and add the workflow to the mlf_core.cfg bump-version configuration. We can then also add this fix later manually here or just sync again later.

**Comment by KevinMenden** [comment]
Aha!
Also add it to the linter 👍 

Thanks @Zethson

**Comment by Zethson** [comment]
> Aha!
> Also add it to the linter +1
> 
> Thanks @Zethson

Think that our linter is smart enough to use the mlf_core.cfg configuration file to verify version integrity. So if we fix this it will automatically also be linted. Am I wrong @Imipenem? Think this is how we implemented this.

**Comment by KevinMenden** [comment]
Of course. I'm not so well versed in `mlf-core linter` black magic :)

**Comment by Imipenem** [comment]
Yes @Zethson ;)

**Comment by Zethson** [comment]
@KevinMenden apparently GA has a limit of 14 GB of mounted space to work with. The latest cudatoolkit etc versions are really fat. This is hard to solve since ideally we would delete useless stuff from the 65GB disk and mount parts of it to finish the build. There are some really hacky solutions out there, but meh.

Maybe try restarting your workflow first. If not we can show you dirty hacks that may have to make it into the templates.

**Comment by KevinMenden** [comment]
Re-running didn't work unfortunately

**Comment by Zethson** [comment]
Sigh. Yeah we are working on this. Looking for a reliable solution...

**Comment by Zethson** [comment]
@KevinMenden could you try the workflow changes in https://github.com/mlf-core/mlf-core/pull/305 here, please?

**Comment by KevinMenden** [comment]
Yes, will try it out this afternoon!

**Comment by KevinMenden** [comment]
Nice, looks like your hack worked @Zethson @Imipenem :+1:

**Review by Zethson — APPROVED** [review]
I might ask you later about the container handling with releases. We have a setup, but I am not sure whether it will work smoothly. Depends on how nicely the Github Container Registry handles multiple tags.

But this PR should be good now.

**Pull request merged**